storage/migration.go

```diff
@@ -0,0 +1,129 @@
+/*
+Copyright © 2020 Red Hat, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package storage
+
+import (
+	"database/sql"
+	"fmt"
+)
+
+// MigrationVersion represents a version of the database.
+type MigrationVersion int
+
+// MigrationStep represents an action performed to either increase
+// or decrease the migration version of the database.
+type MigrationStep func(tx *sql.Tx) error
+
+// Migration type describes a single Migration.
+type Migration struct {
+	StepUp   MigrationStep
+	StepDown MigrationStep
+}
+
+// migrations is a list of migrations that, when applied in their order,
+// create the most recent version of the database from scratch.
+var migrations []Migration
+
+// AddMigration adds a new migration to the list of available migrations.
+func AddMigration(m Migration) {
+	migrations = append(migrations, m)
+}
+
+// InitMigrationInfo creates a table containing migration information in the database.
+func InitMigrationInfo(db *DBStorage) error {
+	tx, err := db.connection.Begin()
+	if err != nil {
+		return err
+	}
+
+	_, err = tx.Exec("CREATE TABLE migration_info (version INTEGER)")
+	if err != nil {
+		tx.Rollback()
+		return err
+	}
+
+	_, err = tx.Exec("INSERT INTO migration_info(version) VALUES(0)")
+	if err != nil {
+		tx.Rollback()
+		return err
+	}
+
+	return tx.Commit()
+}
+
+// GetDBVersion reads the current version of the database from the migration info table.
+func GetDBVersion(db *DBStorage) (MigrationVersion, error) {
+	versionRow := db.connection.QueryRow("SELECT version FROM migration_info")
```

Comment:
prob. add something like LIMIT 1, just to inform that just one record is expected?

Reply:
Agree, to be 100% sure, it will be better to add LIMIT 1

Reply:
My bad, I thought the function would somehow cause an error if there were multiple rows. I wanted to use it to make sure there is exactly 1 row in the table.

If, by some error, the table ended up having multiple rows that were not in sync, we could run into mysterious trouble when using `LIMIT 1`.

I will try to figure out some other way to make sure there's exactly one row in the table.

Reply:
Yup, it might be implemented as a table criteria, if possible (I can imagine one solution which is kinda ugly - have an unique key constraint for another column with constant value).

Reply:
ok this is now implemented properly. TY